{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PinkPigmyPuff/SeatingChart/blob/main/ClassroomGA.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "metadata": {
        "id": "VcPnP3JoMfwj"
      },
      "outputs": [],
      "source": [
        "import random\n",
        "import numpy as np\n",
        "from typing import List, Optional, Callable, Tuple\n",
        "\n",
        "\n",
        "Chromosome = List[int]\n",
        "Population = List[Chromosome]\n",
        "Opinion = List[int]\n",
        "\n",
        "peoplePerTable = 4\n",
        "nextToFriendCost = 10\n",
        "nextToEnemyCost = -.01\n",
        "multipleFriendCost = -.01"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def genChromosome(size: int) -> Chromosome:\n",
        "    return random.sample(range(0, size), size)\n",
        "\n",
        "def genPopulation(popSize: int, chromeosomeSize: int) -> Population:\n",
        "    return [genChromosome(chromeosomeSize) for _ in range(popSize)]\n",
        "\n",
        "def genOpinions(chromosome: Chromosome, opinionMax: int) -> Opinion:\n",
        "  likes = []\n",
        "  dislikes = []\n",
        "  # for every student, randomly generate how many students they have opionions on, and add an empty array to like/dislike\n",
        "  for i in range(0, len(chromosome)):\n",
        "    likes.append([])\n",
        "    dislikes.append([])\n",
        "    likeNum = random.randint(0, opinionMax)\n",
        "    dislikeNum = random.randint(0, opinionMax)\n",
        "    dupelist = chromosome.copy()\n",
        "    dupelist.pop(i)\n",
        "    \n",
        "    # choose random students that the current student likes\n",
        "    for j in range(0, likeNum):\n",
        "      randomStudent = random.randrange(0, len(dupelist))\n",
        "      likes[i].append(dupelist[randomStudent])\n",
        "      dupelist.pop(randomStudent)\n",
        "\n",
        "    # choose random students that the current student dislikes\n",
        "    for j in range(0, dislikeNum):\n",
        "      randomStudent = random.randrange(0, len(dupelist))\n",
        "      dislikes[i].append(dupelist[randomStudent])\n",
        "      dupelist.pop(randomStudent)\n",
        "\n",
        "  return likes, dislikes"
      ],
      "metadata": {
        "id": "klViBf-5MkHU"
      },
      "execution_count": 30,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def alleleFitness(chromosome: Chromosome, allele: int) -> float:\n",
        "      currentStudent = chromosome[allele]\n",
        "      studentsTable = findTable(chromosome, currentStudent)\n",
        "      friendNum = 0\n",
        "      alleleHappiness = 0\n",
        "        # print(f\"current student is {currentStudent}, and they sit at table {studentsTable}\")\n",
        "\n",
        "      # determine the happiness of the currentStudent, based on each person at their table\n",
        "      for tableMate in range(0, peoplePerTable):\n",
        "        currentTableMate = chromosome[studentsTable][tableMate]\n",
        "        if currentTableMate in likes[allele]:\n",
        "          friendNum += 1\n",
        "          if friendNum == 1:\n",
        "            alleleHappiness += nextToFriendCost\n",
        "          if friendNum >= 2:\n",
        "            alleleHappiness += multipleFriendCost\n",
        "\n",
        "        elif currentTableMate in dislikes[allele]:\n",
        "          alleleHappiness += nextToEnemyCost\n",
        "\n",
        "        print(f\"Student {currentStudent} happiness: {alleleHappiness}\")\n",
        "        return alleleHappiness\n",
        "\n",
        "def fitness(chromosome: Chromosome) -> int:\n",
        "    size = len(chromosome)\n",
        "    if size != len(likes) != len(dislikes):\n",
        "        raise ValueError(\"genome and likes/dislikes must be of the same length\")\n",
        "\n",
        "    if size % peoplePerTable != 0:\n",
        "        raise ValueError(\"The length of the chromosome must be divisable by peoplePerTable\")\n",
        "\n",
        "    # determine the cost of the given chromosome\n",
        "    peopleHappiness = [0] * size\n",
        "    chunkedClassroom = chunks(chromosome, peoplePerTable)\n",
        "\n",
        "    # for every student\n",
        "    for student in range(0, len(chromosome)):\n",
        "      peopleHappiness[student] = alleleFitness(chunkedClassroom, student)\n",
        "\n",
        "    return peopleHappiness\n",
        "\n",
        "def populationFitness(population: Population) -> float:\n",
        "    return [average(fitness(chromosome)) for chromosome in population]\n",
        "    \n",
        "# Return n-sized chunks from lst\n",
        "def chunks(lst, n):\n",
        "    x = [lst[i:i + n] for i in range(0, len(lst), n)]\n",
        "    return(x)\n",
        "\n",
        "def relu(x):\n",
        "\t  return max(0.0, x)\n",
        " \n",
        "def average(lst):\n",
        "   return float(sum(lst) / len(lst))"
      ],
      "metadata": {
        "id": "ygCUbNbDWRh2"
      },
      "execution_count": 31,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# finds which table the currentStudent is sitting at\n",
        "def findTable(classroom, student):\n",
        "      if any(student in (match := nested_list) for nested_list in classroom):\n",
        "        return classroom.index(match)"
      ],
      "metadata": {
        "id": "Gq_6zIeNtPiy"
      },
      "execution_count": 32,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Randomly chooses n individuals from the population and returns the fittest one\n",
        "def tournamentSelection(population: Population, tournamentSize: int) -> Population:\n",
        "\t\t\n",
        "\t\ttournament = []\n",
        "\t\ttempPop = population.copy()\n",
        "\t\tfor i in range(0, tournamentSize):\n",
        "\t\t\trandom_id = random.randint(0, len(tempPop)-1)\n",
        "\t\t\ttournament.append(tempPop[random_id])\n",
        "\t\t\ttempPop.pop(i)\n",
        "\n",
        "\t\t#print(f\"tournament pool! {tournament}\")\n",
        "\t\tfittest = chooseHighest(tournament)\n",
        "\t\t#print(f\"fittest! {fittest}\")\n",
        "\t\treturn fittest"
      ],
      "metadata": {
        "id": "stwziiLl_cKw"
      },
      "execution_count": 33,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def roulette_wheel_selection(population: Population, populationFitness) -> Population:\n",
        "    if len(population) != len(populationFitness):\n",
        "        raise ValueError(\"genome and likes/dislikes must be of the same length\")\n",
        "    print(f\"pFit: {populationFitness}\")\n",
        "\n",
        "    # Computes the totallity of the population fitness\n",
        "    sumFitness = sum(populationFitness)\n",
        "    print(f\"sumFittness: {sumFitness}\")\n",
        "    \n",
        "    # Computes for each chromosome the probability \n",
        "    chromosome_probabilities = [classroomHappiness/sumFitness for classroomHappiness in populationFitness]\n",
        "    print(f\"c_prob {chromosome_probabilities}\")\n",
        "\n",
        "    # Selects one chromosome based on the computed probabilities\n",
        "    choice = population[np.random.choice(range(0, len(population)), p=chromosome_probabilities)]\n",
        "    print(f\"chioce_fitness {average(fitness(choice))}\")\n",
        "    return choice"
      ],
      "metadata": {
        "id": "13pyJNJNd1-i"
      },
      "execution_count": 34,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def partiallyMatchedCrossover(ind1, ind2):\n",
        "    #print(f\"ind1: {ind1}\\nind2: {ind2}\")\n",
        "    size = min(len(ind1), len(ind2))\n",
        "    p1, p2 = [0] * size, [0] * size\n",
        "\n",
        "    # Initialize the position of each indices in the individuals\n",
        "    for i in range(size):\n",
        "        p1[ind1[i]] = i\n",
        "        p2[ind2[i]] = i\n",
        "    # Choose crossover points\n",
        "    cxpoint1 = random.randint(0, size)\n",
        "    cxpoint2 = random.randint(0, size - 1)\n",
        "    if cxpoint2 >= cxpoint1:\n",
        "        cxpoint2 += 1\n",
        "    else:  # Swap the two cx points\n",
        "        cxpoint1, cxpoint2 = cxpoint2, cxpoint1\n",
        "\n",
        "    # Apply crossover between cx points\n",
        "    for i in range(cxpoint1, cxpoint2):\n",
        "        # Keep track of the selected values\n",
        "        temp1 = ind1[i]\n",
        "        temp2 = ind2[i]\n",
        "        # Swap the matched value\n",
        "        ind1[i], ind1[p1[temp2]] = temp2, temp1\n",
        "        ind2[i], ind2[p2[temp1]] = temp1, temp2\n",
        "        # Position bookkeeping\n",
        "        p1[temp1], p1[temp2] = p1[temp2], p1[temp1]\n",
        "        p2[temp1], p2[temp2] = p2[temp2], p2[temp1]\n",
        "\n",
        "    #print(f\"FINSIHED, ind1: {ind1}\\n ind2: {ind2}\")\n",
        "    return ind1, ind2"
      ],
      "metadata": {
        "id": "HWx93IfIml3x"
      },
      "execution_count": 35,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def mutate(chromosome: Chromosome) -> Chromosome:\n",
        "    length = len(chromosome)\n",
        "    if length < 2:\n",
        "      print(\"Chromosome is too short to be mutated\")\n",
        "      return chromosome\n",
        "    index1, index2 = random.sample(range(0, length), 2)\n",
        "    chromosome[index1], chromosome[index2] = chromosome[index2], chromosome[index1]\n",
        "    #print(f\"mutated chromosome: {chromosome}\")\n",
        "    return chromosome"
      ],
      "metadata": {
        "id": "0atPm2_xobSl"
      },
      "execution_count": 36,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def chooseHighest(population: Population) -> Chromosome:\n",
        "  #print(f\"pop to choose from: {population}\")\n",
        "  fit = populationFitness(population)\n",
        "  #print(f\"pop fitness: {fit}\")\n",
        "  highestIndex = 0\n",
        "  for i in range(0, len(fit)):\n",
        "      if fit[i] > fit[highestIndex]:\n",
        "          highest = fit[i]\n",
        "          highestIndex = i\n",
        "  #print(f\"chosen: {population[highestIndex]}\")\n",
        "  return population[highestIndex]"
      ],
      "metadata": {
        "id": "7olIOrnRZQQj"
      },
      "execution_count": 37,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def studentInfo(chromosome: Chromosome):\n",
        "  for i in range(len(chromosome)):\n",
        "    chunkedClassroom = chunks(chromosome, 4)\n",
        "          # determine the happiness of the currentStudent, based on each person at their table\n",
        "    studenTable = findTable(chunkedClassroom, i)\n",
        "    print(f\"\\n\\nstudent {i} likes: {likes[i]}, dislikes: {dislikes[i]}, is sitting at table: {studenTable}: \", end=\"\")\n",
        "    for tableMate in range(0, 4):\n",
        "        currentTableMate = chunkedClassroom[studenTable][tableMate]\n",
        "        print(f\"{currentTableMate}\", end=\", \")"
      ],
      "metadata": {
        "id": "qOQJR_zqmnRM"
      },
      "execution_count": 38,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pop = genPopulation(40, 24)\n",
        "likes, dislikes = genOpinions(pop[0], 4)\n",
        "\n",
        "def run_evolution(\n",
        "        generation_limit: int = 100,\n",
        "        population = genPopulation(8, 8), \n",
        "        #likes = genOpinions(population[0], 4),\n",
        "        #dislikes = genOpinions(population[0], 4)\n",
        "        #peoplePerTable = 4,\n",
        "        #nextToFriendCost = 1,\n",
        "        #nextToEnemyCost = -1,\n",
        "        ):\n",
        "  \n",
        "    print(f\"OG population: {population}\")\n",
        "    print(f\"Likes: {likes}\\nDislikes: {dislikes}\")\n",
        "    if len(population) % 2 != 0:\n",
        "        print(\"Population size is odd, removing one chromosome\")\n",
        "        population.pop(0)\n",
        "\n",
        "\n",
        "    for i in range(generation_limit):\n",
        "        nextGeneration = []\n",
        "\n",
        "        # Find and print the fitness of the population, for Debug reasons\n",
        "        popFitness = populationFitness(population)\n",
        "        print(f\"\\n\\nGeneration {i} avg fitness: {average(popFitness)}, {popFitness}\")\n",
        "        print(f\"population: {population}\")\n",
        "\n",
        "        # Create a new parent population, by selecting the most fit members of the population\n",
        "        parents = [tournamentSelection(population, 4) for i in range(len(population))]\n",
        "        print(f\"parents: {parents}\")\n",
        "\n",
        "        # Have one pair of parents breed at a time; then mutate their children, and remove the parents from the breeding pool\n",
        "        while parents:\n",
        "          #print(f\"b0: {parents[0]}\\nb1: {parents[1]}\")\n",
        "          offspring1, offspring2 = partiallyMatchedCrossover(parents[0].copy(), parents[1].copy())\n",
        "          nextGeneration.append(mutate(offspring1.copy()))\n",
        "          nextGeneration.append(mutate(offspring2.copy()))          \n",
        "          parents = parents[2:]\n",
        "\n",
        "        # Set the current population to equal the kids\n",
        "        population = nextGeneration.copy()\n",
        "    \n",
        "    return chooseHighest(population)\n",
        "\n",
        "\n",
        "highest = run_evolution(population=pop)\n",
        "print(highest)\n",
        "print(fitness(highest))\n",
        "print(studentInfo(highest))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 472
        },
        "id": "1pKzKUU0OB06",
        "outputId": "e798a7db-a0fc-4510-a542-fdd0f16cbb96"
      },
      "execution_count": 39,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "OG population: [[17, 12, 16, 18, 8, 6, 2, 15, 22, 11, 20, 9, 14, 5, 10, 13, 0, 4, 21, 1, 19, 3, 23, 7], [1, 12, 10, 16, 23, 3, 9, 13, 17, 4, 21, 22, 20, 18, 14, 0, 7, 2, 8, 15, 11, 19, 6, 5], [8, 13, 17, 11, 16, 19, 15, 7, 20, 5, 4, 9, 1, 22, 21, 14, 3, 23, 2, 6, 0, 10, 18, 12], [14, 21, 0, 18, 16, 2, 17, 4, 19, 23, 9, 8, 12, 1, 3, 7, 11, 15, 13, 20, 10, 5, 22, 6], [20, 3, 21, 10, 7, 4, 2, 22, 9, 6, 8, 1, 5, 14, 15, 23, 13, 17, 19, 12, 18, 11, 0, 16], [5, 11, 20, 23, 13, 10, 16, 4, 19, 2, 0, 1, 3, 21, 9, 22, 14, 8, 6, 17, 15, 12, 18, 7], [23, 10, 21, 5, 8, 17, 14, 2, 1, 13, 11, 3, 0, 19, 4, 15, 6, 20, 22, 12, 9, 16, 18, 7], [11, 8, 16, 4, 6, 14, 22, 13, 3, 9, 7, 12, 15, 21, 10, 0, 2, 1, 19, 20, 23, 17, 18, 5], [23, 5, 19, 8, 22, 20, 6, 15, 7, 17, 3, 10, 13, 21, 9, 16, 14, 0, 1, 12, 2, 18, 11, 4], [18, 4, 20, 13, 2, 7, 5, 12, 0, 16, 1, 6, 3, 8, 14, 11, 15, 10, 19, 21, 23, 9, 22, 17], [8, 16, 19, 2, 21, 15, 9, 10, 20, 7, 0, 18, 6, 12, 5, 22, 23, 3, 17, 14, 11, 4, 1, 13], [16, 17, 14, 19, 3, 23, 22, 5, 13, 12, 18, 15, 10, 11, 9, 0, 21, 2, 1, 20, 7, 4, 6, 8], [20, 18, 17, 22, 4, 3, 15, 23, 9, 13, 19, 14, 2, 21, 7, 5, 11, 10, 16, 12, 8, 0, 6, 1], [23, 22, 15, 4, 0, 6, 5, 8, 16, 3, 19, 2, 12, 7, 13, 1, 14, 20, 9, 11, 21, 10, 18, 17], [16, 21, 22, 5, 4, 12, 19, 18, 15, 0, 7, 2, 8, 3, 11, 6, 1, 13, 20, 14, 9, 10, 17, 23], [4, 13, 11, 2, 6, 9, 1, 7, 21, 15, 19, 23, 10, 14, 17, 22, 12, 18, 8, 16, 20, 0, 3, 5], [19, 1, 0, 2, 15, 3, 23, 4, 14, 17, 5, 22, 13, 10, 6, 18, 20, 21, 12, 7, 9, 11, 16, 8], [9, 8, 12, 21, 0, 2, 3, 15, 10, 16, 6, 17, 5, 11, 18, 22, 20, 14, 1, 19, 13, 4, 23, 7], [16, 14, 6, 7, 10, 20, 1, 11, 5, 22, 15, 8, 12, 4, 9, 0, 2, 21, 18, 23, 13, 17, 3, 19], [21, 15, 2, 8, 6, 7, 17, 1, 13, 20, 0, 10, 19, 23, 11, 14, 3, 12, 5, 9, 22, 18, 16, 4], [4, 2, 0, 14, 23, 16, 7, 5, 22, 1, 20, 15, 3, 13, 18, 17, 11, 21, 19, 6, 10, 8, 12, 9], [6, 18, 14, 9, 11, 13, 12, 15, 4, 10, 17, 19, 8, 3, 7, 2, 16, 23, 21, 5, 22, 0, 20, 1], [11, 1, 8, 19, 23, 0, 18, 5, 20, 7, 22, 12, 10, 14, 15, 21, 16, 13, 6, 17, 4, 3, 9, 2], [23, 11, 13, 22, 0, 20, 7, 5, 8, 18, 17, 2, 4, 3, 15, 12, 9, 16, 14, 10, 21, 1, 6, 19], [13, 5, 8, 2, 17, 1, 0, 21, 7, 6, 18, 15, 9, 22, 16, 4, 10, 19, 14, 23, 12, 3, 11, 20], [13, 7, 3, 8, 21, 14, 5, 16, 11, 10, 12, 0, 15, 2, 4, 6, 9, 22, 20, 18, 23, 1, 17, 19], [4, 19, 16, 5, 3, 20, 8, 11, 2, 23, 14, 22, 21, 9, 13, 10, 12, 0, 6, 7, 15, 1, 17, 18], [18, 19, 17, 13, 21, 12, 8, 14, 1, 0, 3, 11, 20, 10, 16, 4, 5, 2, 9, 22, 6, 7, 23, 15], [11, 5, 1, 19, 13, 18, 8, 2, 20, 17, 7, 9, 4, 21, 12, 3, 14, 16, 22, 10, 6, 0, 15, 23], [12, 17, 13, 7, 21, 16, 8, 15, 20, 10, 22, 19, 0, 1, 6, 18, 3, 14, 2, 11, 4, 23, 9, 5], [20, 6, 9, 22, 23, 8, 2, 13, 18, 12, 21, 7, 11, 14, 4, 16, 3, 1, 5, 17, 19, 15, 10, 0], [4, 2, 23, 15, 5, 20, 16, 10, 14, 22, 18, 1, 7, 13, 19, 17, 6, 3, 8, 12, 9, 0, 21, 11], [22, 23, 20, 17, 5, 1, 4, 21, 7, 2, 19, 9, 18, 11, 14, 13, 12, 16, 0, 8, 15, 3, 10, 6], [5, 20, 12, 3, 8, 1, 6, 23, 13, 2, 7, 17, 18, 19, 16, 22, 15, 0, 11, 14, 9, 4, 10, 21], [7, 9, 22, 0, 3, 10, 21, 5, 1, 11, 20, 8, 6, 4, 13, 2, 15, 14, 12, 19, 23, 16, 17, 18], [7, 21, 5, 17, 11, 8, 9, 20, 6, 18, 23, 3, 12, 4, 22, 2, 19, 15, 13, 0, 1, 10, 14, 16], [12, 4, 15, 19, 2, 17, 11, 6, 5, 21, 8, 20, 18, 0, 13, 14, 22, 7, 23, 3, 16, 1, 10, 9], [17, 9, 1, 18, 10, 12, 11, 16, 3, 20, 8, 5, 2, 13, 15, 21, 22, 23, 7, 0, 4, 19, 14, 6], [19, 10, 4, 2, 9, 6, 12, 20, 3, 0, 13, 14, 22, 1, 23, 8, 17, 18, 15, 11, 7, 5, 21, 16], [13, 3, 5, 9, 11, 4, 19, 14, 22, 18, 20, 7, 15, 6, 8, 1, 2, 0, 12, 16, 10, 21, 23, 17]]\n",
            "Likes: [[20], [], [], [22, 7], [23, 6, 17, 20], [14, 1], [10, 16], [], [16, 0, 1, 17], [2], [9, 16], [], [17, 12], [16], [], [0, 14], [9, 8, 23], [13], [], [17, 14, 0, 11], [18, 9], [19, 13, 18, 16], [16, 21, 7], [12, 15, 3, 5]]\n",
            "Dislikes: [[6, 18, 13, 2], [], [], [14, 9], [16, 22, 10, 5], [], [6, 5, 8], [], [], [17], [4], [6], [15], [9, 7, 8], [7, 4, 6, 1], [16, 17], [], [], [22, 18], [13, 22, 15], [17, 6], [9], [], [21, 18, 4]]\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "TypeError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-39-d5c6acabb494>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     45\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     46\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 47\u001b[0;31m \u001b[0mhighest\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrun_evolution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpopulation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     48\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhighest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     49\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfitness\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhighest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-39-d5c6acabb494>\u001b[0m in \u001b[0;36mrun_evolution\u001b[0;34m(generation_limit, population)\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m         \u001b[0;31m# Find and print the fitness of the population, for Debug reasons\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m         \u001b[0mpopFitness\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpopulationFitness\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpopulation\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"\\n\\nGeneration {i} avg fitness: {average(popFitness)}, {popFitness}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"population: {population}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-31-51a31f167525>\u001b[0m in \u001b[0;36mpopulationFitness\u001b[0;34m(population)\u001b[0m\n\u001b[1;32m     41\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mpopulationFitness\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpopulation\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mPopulation\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 43\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0maverage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfitness\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchromosome\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mchromosome\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mpopulation\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     44\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m \u001b[0;31m# Return n-sized chunks from lst\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-31-51a31f167525>\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     41\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mpopulationFitness\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpopulation\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mPopulation\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 43\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0maverage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfitness\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchromosome\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mchromosome\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mpopulation\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     44\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m \u001b[0;31m# Return n-sized chunks from lst\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-31-51a31f167525>\u001b[0m in \u001b[0;36mfitness\u001b[0;34m(chromosome)\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[0;31m# for every student\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mstudent\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchromosome\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 38\u001b[0;31m       \u001b[0mpeopleHappiness\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstudent\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0malleleFitness\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchunkedClassroom\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstudent\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     39\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mpeopleHappiness\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-31-51a31f167525>\u001b[0m in \u001b[0;36malleleFitness\u001b[0;34m(chromosome, allele)\u001b[0m\n\u001b[1;32m      8\u001b[0m       \u001b[0;31m# determine the happiness of the currentStudent, based on each person at their table\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m       \u001b[0;32mfor\u001b[0m \u001b[0mtableMate\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpeoplePerTable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m         \u001b[0mcurrentTableMate\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mchromosome\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstudentsTable\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtableMate\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcurrentTableMate\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlikes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mallele\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m           \u001b[0mfriendNum\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: list indices must be integers or slices, not NoneType"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Test hundreds of scenarios to determine avg. success\n",
        "def test(n):\n",
        "    pop = genPopulation(40, 20)\n",
        "    likes, dislikes = genOpinions(pop[0], 3)\n",
        "\n",
        "    for i in range(n):\n",
        "        run_evolution(population=pop)"
      ],
      "metadata": {
        "id": "NAnN4hiCqH60"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "1. Encoding scheme\n",
        "- A chromosome is a permutation list\n",
        "- Each index is an integer\n",
        "- Each integer is a student in the class\n",
        "\n",
        "2. Create the initial population\n",
        "- First, a function to create a single chromsome\n",
        "  - inputs chromosomeSize\n",
        "  - outputs a chromosome of randomly shuffled integers\n",
        "\n",
        "7. Testing\n",
        "- A func"
      ],
      "metadata": {
        "id": "d9byY_pyMvft"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Todo:\n",
        "\n",
        "- Add custom classroom layouts, for example with variable table sizes. Could be done w/ an empty, nested list to represent classroom.\n",
        "- Add a CSV -> Data translater\n",
        "- Move out of Google Colab, and into modular scripts (after implementing CSV, etc)\n",
        "- Add a \"special requests\" part to the fitness function, such as needs to be near the front of the class\n",
        "https://github.com/DEAP/deap/blob/master/deap/tools/ "
      ],
      "metadata": {
        "id": "mFdMba1Cmcfa"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import random\n",
        "import warnings\n",
        "\n",
        "try:\n",
        "    from collections.abc import Sequence\n",
        "except ImportError:\n",
        "    from collections import Sequence\n",
        "\n",
        "from itertools import repeat\n",
        "\n",
        "\n",
        "######################################\n",
        "# GA Crossovers                      #\n",
        "######################################\n",
        "\n",
        "\n",
        "def cxOnePoint(ind1, ind2):\n",
        "    \"\"\"Executes a one point crossover on the input :term:`sequence` individuals.\n",
        "    The two individuals are modified in place. The resulting individuals will\n",
        "    respectively have the length of the other.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :returns: A tuple of two individuals.\n",
        "    This function uses the :func:`~random.randint` function from the\n",
        "    python base :mod:`random` module.\n",
        "    \"\"\"\n",
        "    size = min(len(ind1), len(ind2))\n",
        "    cxpoint = random.randint(1, size - 1)\n",
        "    ind1[cxpoint:], ind2[cxpoint:] = ind2[cxpoint:], ind1[cxpoint:]\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxTwoPoint(ind1, ind2):\n",
        "    \"\"\"Executes a two-point crossover on the input :term:`sequence`\n",
        "    individuals. The two individuals are modified in place and both keep\n",
        "    their original length.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :returns: A tuple of two individuals.\n",
        "    This function uses the :func:`~random.randint` function from the Python\n",
        "    base :mod:`random` module.\n",
        "    \"\"\"\n",
        "    size = min(len(ind1), len(ind2))\n",
        "    cxpoint1 = random.randint(1, size)\n",
        "    cxpoint2 = random.randint(1, size - 1)\n",
        "    if cxpoint2 >= cxpoint1:\n",
        "        cxpoint2 += 1\n",
        "    else:  # Swap the two cx points\n",
        "        cxpoint1, cxpoint2 = cxpoint2, cxpoint1\n",
        "\n",
        "    ind1[cxpoint1:cxpoint2], ind2[cxpoint1:cxpoint2] \\\n",
        "        = ind2[cxpoint1:cxpoint2], ind1[cxpoint1:cxpoint2]\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxTwoPoints(ind1, ind2):\n",
        "    \"\"\"\n",
        "    .. deprecated:: 1.0\n",
        "       The function has been renamed.  Use :func:`~deap.tools.cxTwoPoint` instead.\n",
        "    \"\"\"\n",
        "    warnings.warn(\"tools.cxTwoPoints has been renamed. Use cxTwoPoint instead.\",\n",
        "                  FutureWarning)\n",
        "    return cxTwoPoint(ind1, ind2)\n",
        "\n",
        "\n",
        "def cxUniform(ind1, ind2, indpb):\n",
        "    \"\"\"Executes a uniform crossover that modify in place the two\n",
        "    :term:`sequence` individuals. The attributes are swapped according to the\n",
        "    *indpb* probability.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :param indpb: Independent probability for each attribute to be exchanged.\n",
        "    :returns: A tuple of two individuals.\n",
        "    This function uses the :func:`~random.random` function from the python base\n",
        "    :mod:`random` module.\n",
        "    \"\"\"\n",
        "    size = min(len(ind1), len(ind2))\n",
        "    for i in range(size):\n",
        "        if random.random() < indpb:\n",
        "            ind1[i], ind2[i] = ind2[i], ind1[i]\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxPartialyMatched(ind1, ind2):\n",
        "    \"\"\"Executes a partially matched crossover (PMX) on the input individuals.\n",
        "    The two individuals are modified in place. This crossover expects\n",
        "    :term:`sequence` individuals of indices, the result for any other type of\n",
        "    individuals is unpredictable.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :returns: A tuple of two individuals.\n",
        "    Moreover, this crossover generates two children by matching\n",
        "    pairs of values in a certain range of the two parents and swapping the values\n",
        "    of those indexes. For more details see [Goldberg1985]_.\n",
        "    This function uses the :func:`~random.randint` function from the python base\n",
        "    :mod:`random` module.\n",
        "    .. [Goldberg1985] Goldberg and Lingel, \"Alleles, loci, and the traveling\n",
        "       salesman problem\", 1985.\n",
        "    \"\"\"\n",
        "    size = min(len(ind1), len(ind2))\n",
        "    p1, p2 = [0] * size, [0] * size\n",
        "\n",
        "    # Initialize the position of each indices in the individuals\n",
        "    for i in range(size):\n",
        "        p1[ind1[i]] = i\n",
        "        p2[ind2[i]] = i\n",
        "    # Choose crossover points\n",
        "    cxpoint1 = random.randint(0, size)\n",
        "    cxpoint2 = random.randint(0, size - 1)\n",
        "    if cxpoint2 >= cxpoint1:\n",
        "        cxpoint2 += 1\n",
        "    else:  # Swap the two cx points\n",
        "        cxpoint1, cxpoint2 = cxpoint2, cxpoint1\n",
        "\n",
        "    # Apply crossover between cx points\n",
        "    for i in range(cxpoint1, cxpoint2):\n",
        "        # Keep track of the selected values\n",
        "        temp1 = ind1[i]\n",
        "        temp2 = ind2[i]\n",
        "        # Swap the matched value\n",
        "        ind1[i], ind1[p1[temp2]] = temp2, temp1\n",
        "        ind2[i], ind2[p2[temp1]] = temp1, temp2\n",
        "        # Position bookkeeping\n",
        "        p1[temp1], p1[temp2] = p1[temp2], p1[temp1]\n",
        "        p2[temp1], p2[temp2] = p2[temp2], p2[temp1]\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxUniformPartialyMatched(ind1, ind2, indpb):\n",
        "    \"\"\"Executes a uniform partially matched crossover (UPMX) on the input\n",
        "    individuals. The two individuals are modified in place. This crossover\n",
        "    expects :term:`sequence` individuals of indices, the result for any other\n",
        "    type of individuals is unpredictable.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :returns: A tuple of two individuals.\n",
        "    Moreover, this crossover generates two children by matching\n",
        "    pairs of values chosen at random with a probability of *indpb* in the two\n",
        "    parents and swapping the values of those indexes. For more details see\n",
        "    [Cicirello2000]_.\n",
        "    This function uses the :func:`~random.random` and :func:`~random.randint`\n",
        "    functions from the python base :mod:`random` module.\n",
        "    .. [Cicirello2000] Cicirello and Smith, \"Modeling GA performance for\n",
        "       control parameter optimization\", 2000.\n",
        "    \"\"\"\n",
        "    size = min(len(ind1), len(ind2))\n",
        "    p1, p2 = [0] * size, [0] * size\n",
        "\n",
        "    # Initialize the position of each indices in the individuals\n",
        "    for i in range(size):\n",
        "        p1[ind1[i]] = i\n",
        "        p2[ind2[i]] = i\n",
        "\n",
        "    for i in range(size):\n",
        "        if random.random() < indpb:\n",
        "            # Keep track of the selected values\n",
        "            temp1 = ind1[i]\n",
        "            temp2 = ind2[i]\n",
        "            # Swap the matched value\n",
        "            ind1[i], ind1[p1[temp2]] = temp2, temp1\n",
        "            ind2[i], ind2[p2[temp1]] = temp1, temp2\n",
        "            # Position bookkeeping\n",
        "            p1[temp1], p1[temp2] = p1[temp2], p1[temp1]\n",
        "            p2[temp1], p2[temp2] = p2[temp2], p2[temp1]\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxOrdered(ind1, ind2):\n",
        "    \"\"\"Executes an ordered crossover (OX) on the input\n",
        "    individuals. The two individuals are modified in place. This crossover\n",
        "    expects :term:`sequence` individuals of indices, the result for any other\n",
        "    type of individuals is unpredictable.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :returns: A tuple of two individuals.\n",
        "    Moreover, this crossover generates holes in the input\n",
        "    individuals. A hole is created when an attribute of an individual is\n",
        "    between the two crossover points of the other individual. Then it rotates\n",
        "    the element so that all holes are between the crossover points and fills\n",
        "    them with the removed elements in order. For more details see\n",
        "    [Goldberg1989]_.\n",
        "    This function uses the :func:`~random.sample` function from the python base\n",
        "    :mod:`random` module.\n",
        "    .. [Goldberg1989] Goldberg. Genetic algorithms in search,\n",
        "       optimization and machine learning. Addison Wesley, 1989\n",
        "    \"\"\"\n",
        "    size = min(len(ind1), len(ind2))\n",
        "    a, b = random.sample(range(size), 2)\n",
        "    if a > b:\n",
        "        a, b = b, a\n",
        "\n",
        "    holes1, holes2 = [True] * size, [True] * size\n",
        "    for i in range(size):\n",
        "        if i < a or i > b:\n",
        "            holes1[ind2[i]] = False\n",
        "            holes2[ind1[i]] = False\n",
        "\n",
        "    # We must keep the original values somewhere before scrambling everything\n",
        "    temp1, temp2 = ind1, ind2\n",
        "    k1, k2 = b + 1, b + 1\n",
        "    for i in range(size):\n",
        "        if not holes1[temp1[(i + b + 1) % size]]:\n",
        "            ind1[k1 % size] = temp1[(i + b + 1) % size]\n",
        "            k1 += 1\n",
        "\n",
        "        if not holes2[temp2[(i + b + 1) % size]]:\n",
        "            ind2[k2 % size] = temp2[(i + b + 1) % size]\n",
        "            k2 += 1\n",
        "\n",
        "    # Swap the content between a and b (included)\n",
        "    for i in range(a, b + 1):\n",
        "        ind1[i], ind2[i] = ind2[i], ind1[i]\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxBlend(ind1, ind2, alpha):\n",
        "    \"\"\"Executes a blend crossover that modify in-place the input individuals.\n",
        "    The blend crossover expects :term:`sequence` individuals of floating point\n",
        "    numbers.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :param alpha: Extent of the interval in which the new values can be drawn\n",
        "                  for each attribute on both side of the parents' attributes.\n",
        "    :returns: A tuple of two individuals.\n",
        "    This function uses the :func:`~random.random` function from the python base\n",
        "    :mod:`random` module.\n",
        "    \"\"\"\n",
        "    for i, (x1, x2) in enumerate(zip(ind1, ind2)):\n",
        "        gamma = (1. + 2. * alpha) * random.random() - alpha\n",
        "        ind1[i] = (1. - gamma) * x1 + gamma * x2\n",
        "        ind2[i] = gamma * x1 + (1. - gamma) * x2\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxSimulatedBinary(ind1, ind2, eta):\n",
        "    \"\"\"Executes a simulated binary crossover that modify in-place the input\n",
        "    individuals. The simulated binary crossover expects :term:`sequence`\n",
        "    individuals of floating point numbers.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :param eta: Crowding degree of the crossover. A high eta will produce\n",
        "                children resembling to their parents, while a small eta will\n",
        "                produce solutions much more different.\n",
        "    :returns: A tuple of two individuals.\n",
        "    This function uses the :func:`~random.random` function from the python base\n",
        "    :mod:`random` module.\n",
        "    \"\"\"\n",
        "    for i, (x1, x2) in enumerate(zip(ind1, ind2)):\n",
        "        rand = random.random()\n",
        "        if rand <= 0.5:\n",
        "            beta = 2. * rand\n",
        "        else:\n",
        "            beta = 1. / (2. * (1. - rand))\n",
        "        beta **= 1. / (eta + 1.)\n",
        "        ind1[i] = 0.5 * (((1 + beta) * x1) + ((1 - beta) * x2))\n",
        "        ind2[i] = 0.5 * (((1 - beta) * x1) + ((1 + beta) * x2))\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxSimulatedBinaryBounded(ind1, ind2, eta, low, up):\n",
        "    \"\"\"Executes a simulated binary crossover that modify in-place the input\n",
        "    individuals. The simulated binary crossover expects :term:`sequence`\n",
        "    individuals of floating point numbers.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :param eta: Crowding degree of the crossover. A high eta will produce\n",
        "                children resembling to their parents, while a small eta will\n",
        "                produce solutions much more different.\n",
        "    :param low: A value or a :term:`python:sequence` of values that is the lower\n",
        "                bound of the search space.\n",
        "    :param up: A value or a :term:`python:sequence` of values that is the upper\n",
        "               bound of the search space.\n",
        "    :returns: A tuple of two individuals.\n",
        "    This function uses the :func:`~random.random` function from the python base\n",
        "    :mod:`random` module.\n",
        "    .. note::\n",
        "       This implementation is similar to the one implemented in the\n",
        "       original NSGA-II C code presented by Deb.\n",
        "    \"\"\"\n",
        "    size = min(len(ind1), len(ind2))\n",
        "    if not isinstance(low, Sequence):\n",
        "        low = repeat(low, size)\n",
        "    elif len(low) < size:\n",
        "        raise IndexError(\"low must be at least the size of the shorter individual: %d < %d\" % (len(low), size))\n",
        "    if not isinstance(up, Sequence):\n",
        "        up = repeat(up, size)\n",
        "    elif len(up) < size:\n",
        "        raise IndexError(\"up must be at least the size of the shorter individual: %d < %d\" % (len(up), size))\n",
        "\n",
        "    for i, xl, xu in zip(range(size), low, up):\n",
        "        if random.random() <= 0.5:\n",
        "            # This epsilon should probably be changed for 0 since\n",
        "            # floating point arithmetic in Python is safer\n",
        "            if abs(ind1[i] - ind2[i]) > 1e-14:\n",
        "                x1 = min(ind1[i], ind2[i])\n",
        "                x2 = max(ind1[i], ind2[i])\n",
        "                rand = random.random()\n",
        "\n",
        "                beta = 1.0 + (2.0 * (x1 - xl) / (x2 - x1))\n",
        "                alpha = 2.0 - beta ** -(eta + 1)\n",
        "                if rand <= 1.0 / alpha:\n",
        "                    beta_q = (rand * alpha) ** (1.0 / (eta + 1))\n",
        "                else:\n",
        "                    beta_q = (1.0 / (2.0 - rand * alpha)) ** (1.0 / (eta + 1))\n",
        "\n",
        "                c1 = 0.5 * (x1 + x2 - beta_q * (x2 - x1))\n",
        "\n",
        "                beta = 1.0 + (2.0 * (xu - x2) / (x2 - x1))\n",
        "                alpha = 2.0 - beta ** -(eta + 1)\n",
        "                if rand <= 1.0 / alpha:\n",
        "                    beta_q = (rand * alpha) ** (1.0 / (eta + 1))\n",
        "                else:\n",
        "                    beta_q = (1.0 / (2.0 - rand * alpha)) ** (1.0 / (eta + 1))\n",
        "                c2 = 0.5 * (x1 + x2 + beta_q * (x2 - x1))\n",
        "\n",
        "                c1 = min(max(c1, xl), xu)\n",
        "                c2 = min(max(c2, xl), xu)\n",
        "\n",
        "                if random.random() <= 0.5:\n",
        "                    ind1[i] = c2\n",
        "                    ind2[i] = c1\n",
        "                else:\n",
        "                    ind1[i] = c1\n",
        "                    ind2[i] = c2\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "######################################\n",
        "# Messy Crossovers                   #\n",
        "######################################\n",
        "\n",
        "def cxMessyOnePoint(ind1, ind2):\n",
        "    \"\"\"Executes a one point crossover on :term:`sequence` individual.\n",
        "    The crossover will in most cases change the individuals size. The two\n",
        "    individuals are modified in place.\n",
        "    :param ind1: The first individual participating in the crossover.\n",
        "    :param ind2: The second individual participating in the crossover.\n",
        "    :returns: A tuple of two individuals.\n",
        "    This function uses the :func:`~random.randint` function from the python base\n",
        "    :mod:`random` module.\n",
        "    \"\"\"\n",
        "    cxpoint1 = random.randint(0, len(ind1))\n",
        "    cxpoint2 = random.randint(0, len(ind2))\n",
        "    ind1[cxpoint1:], ind2[cxpoint2:] = ind2[cxpoint2:], ind1[cxpoint1:]\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "######################################\n",
        "# ES Crossovers                      #\n",
        "######################################\n",
        "\n",
        "def cxESBlend(ind1, ind2, alpha):\n",
        "    \"\"\"Executes a blend crossover on both, the individual and the strategy. The\n",
        "    individuals shall be a :term:`sequence` and must have a :term:`sequence`\n",
        "    :attr:`strategy` attribute. Adjustment of the minimal strategy shall be done\n",
        "    after the call to this function, consider using a decorator.\n",
        "    :param ind1: The first evolution strategy participating in the crossover.\n",
        "    :param ind2: The second evolution strategy participating in the crossover.\n",
        "    :param alpha: Extent of the interval in which the new values can be drawn\n",
        "                  for each attribute on both side of the parents' attributes.\n",
        "    :returns: A tuple of two evolution strategies.\n",
        "    This function uses the :func:`~random.random` function from the python base\n",
        "    :mod:`random` module.\n",
        "    \"\"\"\n",
        "    for i, (x1, s1, x2, s2) in enumerate(zip(ind1, ind1.strategy,\n",
        "                                             ind2, ind2.strategy)):\n",
        "        # Blend the values\n",
        "        gamma = (1. + 2. * alpha) * random.random() - alpha\n",
        "        ind1[i] = (1. - gamma) * x1 + gamma * x2\n",
        "        ind2[i] = gamma * x1 + (1. - gamma) * x2\n",
        "        # Blend the strategies\n",
        "        gamma = (1. + 2. * alpha) * random.random() - alpha\n",
        "        ind1.strategy[i] = (1. - gamma) * s1 + gamma * s2\n",
        "        ind2.strategy[i] = gamma * s1 + (1. - gamma) * s2\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxESTwoPoint(ind1, ind2):\n",
        "    \"\"\"Executes a classical two points crossover on both the individuals and their\n",
        "    strategy. The individuals shall be a :term:`sequence` and must have a\n",
        "    :term:`sequence` :attr:`strategy` attribute. The crossover points for the\n",
        "    individual and the strategy are the same.\n",
        "    :param ind1: The first evolution strategy participating in the crossover.\n",
        "    :param ind2: The second evolution strategy participating in the crossover.\n",
        "    :returns: A tuple of two evolution strategies.\n",
        "    This function uses the :func:`~random.randint` function from the python base\n",
        "    :mod:`random` module.\n",
        "    \"\"\"\n",
        "    size = min(len(ind1), len(ind2))\n",
        "\n",
        "    pt1 = random.randint(1, size)\n",
        "    pt2 = random.randint(1, size - 1)\n",
        "    if pt2 >= pt1:\n",
        "        pt2 += 1\n",
        "    else:  # Swap the two cx points\n",
        "        pt1, pt2 = pt2, pt1\n",
        "\n",
        "    ind1[pt1:pt2], ind2[pt1:pt2] = ind2[pt1:pt2], ind1[pt1:pt2]\n",
        "    ind1.strategy[pt1:pt2], ind2.strategy[pt1:pt2] = \\\n",
        "        ind2.strategy[pt1:pt2], ind1.strategy[pt1:pt2]\n",
        "\n",
        "    return ind1, ind2\n",
        "\n",
        "\n",
        "def cxESTwoPoints(ind1, ind2):\n",
        "    \"\"\"\n",
        "    .. deprecated:: 1.0\n",
        "        The function has been renamed. Use :func:`cxESTwoPoint` instead.\n",
        "    \"\"\"\n",
        "    return cxESTwoPoint(ind1, ind2)\n",
        "\n",
        "\n",
        "# List of exported function names.\n",
        "__all__ = ['cxOnePoint', 'cxTwoPoint', 'cxUniform', 'cxPartialyMatched',\n",
        "           'cxUniformPartialyMatched', 'cxOrdered', 'cxBlend',\n",
        "           'cxSimulatedBinary', 'cxSimulatedBinaryBounded', 'cxMessyOnePoint',\n",
        "           'cxESBlend', 'cxESTwoPoint']"
      ],
      "metadata": {
        "id": "2wWI-QohrPNY"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}